File: events/meetup_dotnet_architects_2015/Fun with Functional Programming/F#/Lifting.fs
//Sample provided by Fabio Galuppo 
//February 2015 

module LiftingProgram

let lift (f : 'a -> 'b) : ('a option -> 'b option) =
    fun x ->
        match x with
        | Some(value) -> Some (f(value))
        | None -> None

let xs = [1..10]
let ys = List.map (fun x -> Some x) xs

let int2float (x : int32) : float32 = float32 x

let test1() =
    let zs = xs |> List.map (int2float)
    printfn "%A" zs
    let ws = ys |> List.map (lift int2float)
    printfn "%A" ws

File: events/meetup_dotnet_architects_2015/Fun with Functional Programming/C#/Lifting.cs
//Sample provided by Fabio Galuppo 
//February 2015 

using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharpFPIntro
{
    static class LiftingProgram
    {
        static float IntToFloat(int x)
        {
            return Convert.ToSingle(x);
        }

        static Func<Nullable<TInput>, Nullable<TOutput>> Lift<TInput, TOutput>(Func<TInput, TOutput> f)
            where TInput : struct
            where TOutput : struct
        {
            return x =>
            {
                if (x.HasValue)
                    return new Nullable<TOutput>(f(x.Value));
                return new Nullable<TOutput>();
            };
        }

        public static void Test1()
        {
            var xs = Enumerable.Range(1, 10);

            IEnumerable<float> ys = xs.Select(x => 10.0f * x).ToArray();

            IEnumerable<float> zs = xs.Select(IntToFloat).ToArray();

            IEnumerable<Nullable<int>> ws = Enumerable.Range(1, 10).Select(x => new Nullable<int>(x));

            //IEnumerable<Nullable<float>> vs = ws.Select(IntToFloat).ToArray(); //?
            IEnumerable<Nullable<float>> vs = ws.Select(Lift<int, float>(IntToFloat)).ToArray();
            foreach (var v in vs)
                if (v.HasValue) Console.Write("{0:0.0} ", v);
                else Console.Write("Null ");
            Console.WriteLine();
        }
    }
}
